File: 1semester/dz6/dz6_2/dz6_2/stack.cpp
#include "stack.h"
#include <stdio.h>
#include <string.h>
#include <locale.h>

struct StackElement
{
	char value;
	StackElement *next;
};

struct Stack
{
	StackElement *head;
};

Stack *createStack()
{
	Stack *result = new Stack;
	result->head = NULL;
	return result;
}

void push(Stack *stack, char value)
{
	StackElement *newHead = new StackElement;
	newHead->next = stack->head;
	newHead->value = value;
	stack->head = newHead;
}

char pop(Stack *stack)
{
	if (stack->head != NULL)
	{
		char result = stack->head->value;
		StackElement *temp = stack->head;
		stack->head = stack->head->next;
		delete temp;
		return result;
	}
	else
	{
		printf("Стек пустой!");
		return '0';
	}
}

void printStack(Stack *stack)
{
	while (stack->head != NULL)
	{
		printf("%c\n", stack->head->value);
		pop(stack);
	}
}

void delStack(Stack *stack)
{
	StackElement *temp = stack->head;
	while (temp != NULL)
	{
		stack->head = stack->head->next;
		delete temp;
		temp = stack->head;
	}
	delete stack;
}

char headValue(Stack *stack)
{
	return stack->head->value;
}

File: 2semester/Homework7/hw7.task1/Stack/src/stack/Stack.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package stack;

/**
 *
 * @author Miracle
 */
public class Stack <Type> {
    /**
     * push value to the stack
     *
     * @param value value, which you want to push
     */
    public void push(Type valueForAdd) {
        StackElement temp = new StackElement(valueForAdd, head);
        head = temp;
    }

    /**
     * pop an element from the stack
     *
     * @return value of the top element
     */
    public Type pop() throws EmptyStack {
        if (isEmpty()) {
            throw new EmptyStack();
        } else {
        Type answer = head.value;
        head = head.next;
        return answer;
        }
    }

    /**
     * to return value of the top element
     *
     * @return value of the top element
     */
    public Type top() throws EmptyStack {
        if (isEmpty()) {
            throw new EmptyStack();           
        } else {
            return head.value;
        }
    }

    /**
     * checks that the stack is empty or not
     *
     * @return true if stack is empty else false
     */
    public boolean isEmpty() {
        return head == null;
    }

    /**
     * element of the stack
     */
    private class StackElement {

        public StackElement(Type valueForAdd, StackElement next) {
            this.value = valueForAdd;
            this.next = next;
        }
        private Type value;
        private StackElement next;
    }
    /**
     * the head of the stack
     */
    private StackElement head;
    /**
     * @param args the command line arguments
     */
    public static void main(String[] args) {
        // TODO code application logic here
    }
}
